File: demos/aspnet-react-feequote/src/FeeQuote.Web/Controllers/Api/QuotesController.cs
using System.ComponentModel.DataAnnotations;
using FeeQuote.Web.Domain.Config;
using FeeQuote.Web.Domain.Models;
using FeeQuote.Web.Domain.Services;
using Microsoft.AspNetCore.Mvc;

namespace FeeQuote.Web.Controllers.Api;

/// <summary>Request body for POST /api/quotes.</summary>
public sealed record QuoteRequest(
    [Range(1, long.MaxValue)] long AmountCents,
    [Required] string Currency,
    [Required] string OriginCountry,
    [Required] string DestinationCountry,
    string? Channel);

/// <summary>Response body for POST /api/quotes. Field names are the public contract.</summary>
public sealed record QuoteResponse(
    string TransferId,
    long AmountCents,
    string Currency,
    long FeeCents,
    long TotalCents,
    IReadOnlyList<BreakdownLine> Breakdown);

[ApiController]
[Route("api/quotes")]
[Produces("application/json")]
public sealed class QuotesController : ControllerBase
{
    private readonly TransferService _transfers;

    public QuotesController(TransferService transfers)
    {
        _transfers = transfers;
    }

    [HttpPost]
    [ProducesResponseType<QuoteResponse>(StatusCodes.Status200OK)]
    [ProducesResponseType<ValidationProblemDetails>(StatusCodes.Status400BadRequest)]
    public ActionResult<QuoteResponse> Create([FromBody] QuoteRequest request)
    {
        ArgumentNullException.ThrowIfNull(request);

        var transfer = new Transfer(
            Id: Guid.NewGuid().ToString("N"),
            AmountCents: request.AmountCents,
            Currency: request.Currency.ToUpperInvariant(),
            OriginCountry: request.OriginCountry.ToUpperInvariant(),
            DestinationCountry: request.DestinationCountry.ToUpperInvariant(),
            Channel: string.IsNullOrWhiteSpace(request.Channel) ? Settings.DefaultChannel : request.Channel);

        try
        {
            var quote = _transfers.Quote(transfer);
            return Ok(new QuoteResponse(
                quote.TransferId,
                transfer.AmountCents,
                transfer.Currency,
                quote.FeeCents,
                quote.TotalCents,
                quote.Breakdown));
        }
        catch (ArgumentException ex)
        {
            ModelState.AddModelError("transfer", ex.Message);
            return ValidationProblem(ModelState);
        }
    }
}

File: demos/java-feequote/src/main/java/com/feequote/web/QuotesController.java
package com.feequote.web;

import com.feequote.config.Settings;
import com.feequote.models.BreakdownLine;
import com.feequote.models.Quote;
import com.feequote.models.Transfer;
import com.feequote.services.TransferService;
import com.feequote.services.TransferValidationException;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.Positive;
import java.util.List;
import java.util.Locale;
import java.util.UUID;
import org.springframework.http.HttpStatus;
import org.springframework.http.ProblemDetail;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

/** The quoting endpoint. Field names on the request and the response are the public contract. */
@RestController
@RequestMapping("/api/quotes")
public class QuotesController {

    /** Request body for POST /api/quotes. */
    public record QuoteRequest(
            @Positive long amountCents,
            @NotBlank String currency,
            @NotBlank String originCountry,
            @NotBlank String destinationCountry,
            String channel) {
    }

    /** Response body for POST /api/quotes. */
    public record QuoteResponse(
            String transferId,
            long amountCents,
            String currency,
            long feeCents,
            long totalCents,
            List<BreakdownLine> breakdown) {
    }

    private final TransferService transfers;

    public QuotesController(TransferService transfers) {
        this.transfers = transfers;
    }

    /** Prices a transfer and returns the quote with its fee breakdown. */
    @PostMapping
    public QuoteResponse create(@Valid @RequestBody QuoteRequest request) {
        Transfer transfer = new Transfer(
                UUID.randomUUID().toString().replace("-", ""),
                request.amountCents(),
                upper(request.currency()),
                upper(request.originCountry()),
                upper(request.destinationCountry()),
                blankToDefault(request.channel()));

        Quote quote = transfers.quote(transfer);
        return new QuoteResponse(
                quote.transferId(),
                transfer.amountCents(),
                transfer.currency(),
                quote.feeCents(),
                quote.totalCents(),
                quote.breakdown());
    }

    /** A transfer the service cannot price is a bad request, not a server error. */
    @ExceptionHandler(TransferValidationException.class)
    public ProblemDetail onValidationFailure(TransferValidationException failure) {
        ProblemDetail problem = ProblemDetail.forStatus(HttpStatus.BAD_REQUEST);
        problem.setTitle("The transfer cannot be quoted");
        problem.setDetail(failure.getMessage());
        return problem;
    }

    private static String upper(String value) {
        return value.toUpperCase(Locale.ROOT);
    }

    private static String blankToDefault(String channel) {
        if (channel == null || channel.isBlank()) {
            return Settings.DEFAULT_CHANNEL;
        }
        return channel;
    }
}
